src/utils/common/cyclic_buffer.h

```diff
@@ -0,0 +1,361 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef _NIXL_CYCLIC_BUFFER_H
+#define _NIXL_CYCLIC_BUFFER_H
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <atomic>
+#include <cstring>
+#include <stdexcept>
+#include <iostream>
+#include <string>
+
+#include "common/nixl_log.h"
+#include "util.h"
+
+template<typename T> class SharedRingBuffer {
+private:
+    struct BufferHeader {
+        std::atomic<size_t> write_pos{0};
+        std::atomic<size_t> read_pos{0};
+        std::atomic<uint32_t> version{0};
+        uint32_t expected_version{0};
+        size_t capacity;
+        size_t mask;
+
+        BufferHeader(size_t size) : capacity(size), mask(size - 1) {
+            // Ensure size is power of 2
+            if ((size & (size - 1)) != 0) {
+                size = nextPowerOf2(size);
+                NIXL_WARN << "Size " << size << " is not power of 2, rounding up to: " << size;
+            }
+            static_assert(std::is_trivially_copyable<T>::value,
+                          "T must be trivially copyable for shared memory");
+        }
+    };
+
+    BufferHeader *header_;
+    T *data_;
+    int file_fd_;
+    bool is_creator_;
+    std::string file_path_;
+    bool initialized_;
+    size_t buffer_size_;
+
+    size_t
+    getTotalSize() const {
+        return sizeof(BufferHeader) + sizeof(T) * buffer_size_;
+    }
+
+public:
+    SharedRingBuffer()
+        : header_(nullptr),
+          data_(nullptr),
+          file_fd_(-1),
+          is_creator_(false),
+          file_path_(""),
+          initialized_(false),
+          buffer_size_(0) {}
+
+    SharedRingBuffer(const char *name, size_t size, bool create, uint32_t version = 1)
+        : file_fd_(-1),
+          is_creator_(create),
+          file_path_(name),
+          initialized_(false),
+          buffer_size_(size) {
+        initialize(name, size, create, version);
+    }
+
+    // Constructor for reading existing buffer (size will be read from header)
+    SharedRingBuffer(const char *name, uint32_t version = 1)
+        : file_fd_(-1),
+          is_creator_(false),
+          file_path_(name),
+          initialized_(false),
+          buffer_size_(0) {
+        initialize(name, 0, false, version);
+    }
+
+    void
+    initialize(const char *name, size_t size = 0, bool create = true, uint32_t version = 1) {
```

Comment:
Let's only provide default values where it makes sense. Here size==0 just results an error on line 102.